src/main/kotlin/kweb/WebBrowser.kt

```diff
@@ -59,37 +65,79 @@ class WebBrowser(private val sessionId: String, val httpRequestInfo: HttpRequest
+            params.deleteCharAt(params.lastIndex)//delete the last trailing comma
+        }
+        return JSFunction(js, params.toString())
     }
 
-    fun executeWithCallback(js: String, callbackId: Int, callback: (Any) -> Unit) {
-        kweb.executeWithCallback(sessionId, js, callbackId, callback)
+    fun callJs(js: String, vararg args: Any?) {
+        cachedFunctions[js]?.let {
+            val server2ClientMessage = Server2ClientMessage(yourId = sessionId, jsId = it, arguments = listOf(*args), js = js)
+            kweb.callJs(server2ClientMessage, js)
+        } ?: run {
+            val rng = Random()
+            val cacheId = rng.nextInt()
+            val func = getJsFunction(js)
+            //we add the user's unmodified js as a key and the cacheId as it's value in the hashmap
+            cachedFunctions[js] = cacheId
+            //we send the modified js to the client to be cached there.
+            //we don't cache the modified js on the server, because then we'd have to modify JS on the server, everytime we want to check the server's cache
+            val server2ClientMessage = Server2ClientMessage(yourId = sessionId, jsId = cacheId, js = func.js,
+                parameters = func.params, arguments = listOf(*args))
+            kweb.callJs(server2ClientMessage, js)
+        }
+    }
+
+    fun callJsWithCallback(js: String, callbackId: Int, callback: (Any) -> Unit, vararg args: Any?) {
+        cachedFunctions[js]?.let {
+            val server2ClientMessage = Server2ClientMessage(yourId = sessionId, jsId = it, arguments = listOf(*args),
+            callbackId = callbackId, js = js)
+            kweb.callJsWithCallback(server2ClientMessage, js, callback)
+        } ?: run {
+            val rng = Random()
+            val cacheId = rng.nextInt()
+            val func = getJsFunction(js)
+            //we add the user's unmodified js as a key and the cacheId as it's value in the hashmap
+            cachedFunctions[js] = cacheId
+            //we send the modified js to the client to be cached there.
+            //we don't cache the modified js on the server, because then we'd have to modify JS on the server, everytime we want to check the server's cache
+            val server2ClientMessage = Server2ClientMessage(yourId = sessionId, jsId = cacheId, js = func.js,
+                    parameters = func.params, arguments = listOf(*args), callbackId = callbackId)
+            kweb.callJsWithCallback(server2ClientMessage, js, callback)
+        }
     }
 
     fun removeCallback(callbackId: Int) {
         kweb.removeCallback(sessionId, callbackId)
     }
 
-    fun evaluate(js: String): CompletableFuture<Any> {
+    /*TODO
+    I'm not 100% sure what to do with this function. I think it'd be too complicated to merge it into
+    one of the callJs functions.
+    Maybe it should just be named something like 'callJsWithResult' or something.*/
+    fun evaluate(js: String, vararg args: Any?): CompletableFuture<Any> {
```

Comment:
I remember talking about this. I forgot you mentioned that. I just tried rewriting it as a suspend function. I then went to the first place it's called in CookieReceiver, and was getting an error. Also, that call is written to handle a CompletableFuture.
This will probably have to work as a suspend function. But, I'm gonna have to play with it for a little while, and probably rewrite some code in the places where it is currently called.